### ISSUE TITLE
links should be urlencoded when using a dereferer

### ISSUE THREAD

**Issue by jcsaaddupuy** [shaarli/Shaarli#328]
Currently, if you use a dereferer in shaarli, there is an issue with links containing a fragment (the part after #).

As the url is not encoded, the browser parse the fragment and does not send it to to the dereferer.

Example with dereferer.org and zerobin :

With url not encoded (we are redirected to https://www.zerobin.net/?b82c165d5e558015) :

http://dereferer.org/?https://www.zerobin.net/?b82c165d5e558015#pxU6FIkc+J/LJ/Wj7Sr+2ZCkJZwiU+kWCWoSlmVqBdQ=
 

With url encoded, we are correctly redirected :

http://dereferer.org/?https%3A//www.zerobin.net/%3Fb82c165d5e558015%23pxU6FIkc%2BJ/LJ/Wj7Sr%2B2ZCkJZwiU%2BkWCWoSlmVqBdQ%3D

**Comment by ArthurHoaro** [comment]
You're saving an invalid link in Shaarli (`http://dereferer.org/?https://www.zerobin.net/?b82c165d5e558015#pxU6FIkc+J/LJ/Wj7Sr+2ZCkJZwiU+kWCWoSlmVqBdQ=`). It's obiously not working.

Shaarli's giving you what you saved and is not supposed to know what you're doing with your link (plus it would break any other link).

I suggest that you store a valid link, or contact dereferer.org to make them handle unencoded URIs.

**Comment by jcsaaddupuy** [comment]
> I suggest that you store a valid link,

I think there is a little misunderstanding here.

Both links are totally valid.

- the first one is the link that shaarli will generate with http://dereferer.org/? configured as the dereferer, but as the url  is not url encoded, the fragment (the part after #) won't be sent by the browser to the dereferer.org server (that's how fragments works). You see an error when clicking on it because the fragment containing the decryption key is droped by the browser.

- the second link is a the very same url, but urlencoded. The web browser won't drop the fragment in the url when sending it to dereferer.org.


You cant test it with any other url having a fragment.
Let says we use the original url https://en.wikipedia.org/wiki/Brazil_%281985_film%29#Plot (point to the 'Plot' section of the film Brazil - great movie)
If you save this link in shaarli,  shaarli wil generate the url
http://dereferer.org/?https://en.wikipedia.org/wiki/Brazil_%281985_film%29#Plot 

As you can see,  the '#' char is not urlencoded and will never be sent by the browser to the dereferer server. You 'll be redirected to the top of the wikipedia article.

If you urlencode the url, shaarli will generate the following :
http://dereferer.org/?https%3A//en.wikipedia.org/wiki/Brazil_%25281985_film%2529%23Plot

The '#' is urlencoded. You'll be redirected to the 'Plot' section on the wikipedia page, as with the original link.

In short : if you save a link containing a fragment in shaarli while using a dereferer, you will loose whatever is after the fragment while being redirected.


>  or contact dereferer.org to make them handle unencoded URIs.

They can't. Any dereferer won't ever receive the fragment if not urlencoded.
I would rather prefer that shaarli use their service properly :) (and I may provide a PR later for this issue.)

**Comment by nodiscc** [comment]
I think the behaviour described by @jcsaaddupuy is actually a _Shaarli_ bug.

At the very least, the `#` character should be urlencoded passing the URL to the no-referer service (else it has zero chances to work). Wikipedia anchors and the zerobin description key are good examples.

Anyone agree?

**Comment by ArthurHoaro** [comment]
I might be missing something but... if you save not working links in Shaarli, you end up with not working links. I don't see where the bug stands.

Plus, your solution would turn a perfectly working link into : https://en.m.wikipedia.org/wiki/Brazil_%25281985_film%2529%23Plot

Unless dereferer.org is related to Shaarli in any way? 
-- 
Arthur Hoaro
http://hoa.ro
- Envoi mobile -

**Comment by jcsaaddupuy** [comment]
The link saved in shaarli is perfectly working.
The link generated by shaarli is not.

> Plus, your solution would turn a perfectly working link into : https://en.m.wikipedia.org/wiki/Brazil_%25281985_film%2529%23Plot

Nope. As the issue title said, this should only be done when using a redirector.

This is not a dereferer.org specific issue. You'll end up with the very similar result for any redirector you use.

**Comment by ArthurHoaro** [comment]
Aaaaah! I totally forgot about the "Redirector" feature in Shaarli's settings. Sorry about that.

Yes, we probably want to `urlencode()` links in that case.
